### ISSUE TITLE
sometime sc.pp.filter_genes_dispersion return the  gene_subset with the dimension of  n_top_genes-1 

### ISSUE THREAD

**Issue by yuxiaokang-source** [scverse/scanpy#1457]
- [ ] I have checked that this issue has not already been reported.
- [ ] I have confirmed this bug exists on the latest version of scanpy.
- [ ] (optional) I have confirmed this bug exists on the master branch of scanpy.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

### Minimal code sample (that we can copy&paste without having any data)
```
the simple simulation dataset(filter_gene_dis.h5,which is the adata in the simulation code) is
AnnData object with n_obs × n_vars = 3 × 50
[[ 8. 10.  7. 11. 14. 12. 11.  9. 14. 11.  8.  2.  8. 10.  7. 12. 11. 12.
  12. 10.  3. 11. 13.  7.  8. 11. 14.  9. 11.  9.  5. 13.  8. 13.  9. 15.
  11.  8.  7.  7.  5. 12.  9. 12. 11.  8. 11.  6. 10. 11.]
 [19. 22. 19. 23. 20. 16. 13. 26. 20. 29. 22. 16. 19. 22. 24. 20. 19. 15.
  17. 25. 23. 19. 18. 18. 24. 18. 25. 22. 25. 16. 25. 23. 27. 22. 14. 21.
  24. 23. 16. 15. 14. 27. 23. 24. 21. 27. 17. 20. 20. 12.]
 [27. 31. 32. 30. 29. 31. 24. 29. 29. 33. 29. 29. 26. 38. 27. 32. 21. 24.
  28. 27. 25. 19. 28. 24. 23. 23. 30. 39. 29. 42. 34. 28. 25. 26. 27. 32.
  28. 35. 34. 26. 27. 22. 24. 42. 30. 32. 29. 28. 29. 34.]]```
```

```python
import scanpy as sc
import logging
sc.settings.verbosity = 3    
adata=sc.read("filter_gene_dis.h5")#a simulation dataset
#when n_top_genes=10,it returns 3*10
#when n_top_genes=20,it returns 3*19
sc.pp.filter_genes_dispersion(adata,n_top_genes=20)
print(adata)
```

```pytb
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
AnnData object with n_obs × n_vars = 3 × 19
    var: 'means', 'dispersions', 'dispersions_norm'
```

#### Versions

<details>
scanpy==1.6.0 anndata==0.7.4 umap==0.4.6 numpy==1.18.5 scipy==1.5.0 pandas==1.0.5 scikit-learn==0.23.1 statsmodels==0.11.1 python-igraph==0.8.2 louvain==0.7.0 leidenalg==0.8.1
</details>

**Comment by yuxiaokang-source** [comment]
in the source file(highly_variable_genes.py)
```
    dispersion_norm = df['dispersion_norm'].values.astype('float32')
    if n_top_genes is not None:
        dispersion_norm = dispersion_norm[~np.isnan(dispersion_norm)]
        dispersion_norm[::-1].sort()  # interestingly, np.argpartition is slightly slower
        disp_cut_off = dispersion_norm[n_top_genes-1]
        gene_subset = df['dispersion_norm'].values >= disp_cut_off
        logg.debug(
            f'the {n_top_genes} top genes correspond to a '
            f'normalized dispersion cutoff of {disp_cut_off}'
        )
``` 
the ` df['dispersion_norm']` is the type of `np.float64`,but the `dispersion_norm` is `np.float32`,so the `disp_cut_off` is `np.float32`,but in line 
`gene_subset = df['dispersion_norm'].values >= disp_cut_off` , 
It means the array of `float64` will compare with `float32`, it is not accurate sometimes.
in my simulation code, the dubug result is follows
```
disp_cut_off
0.5938217

type(disp_cut_off)
<class 'numpy.float32'>

df['dispersion_norm'][44]
0.593821696856851

type(df['dispersion_norm'][44])
<class 'numpy.float64'>

df['dispersion_norm'][44]>=disp_cut_off
False
```

**Comment by LuckyMD** [comment]
There is already a PR to fix this #1456. This should be merged into master soon.
